providers/common/include/internal/digestcommon.h

```diff
@@ -0,0 +1,114 @@
+{ OSSL_FUNC_DIGEST_GETTABLE_PARAMS,                                            \
+  (void (*)(void))digest_default_gettable_params }
+
+# define OSSL_DISPATCH_FUNC_DIGEST_COMMON(name, init, upd)                     \
+const OSSL_DISPATCH name##_functions[] = {                                     \
+    { OSSL_FUNC_DIGEST_NEWCTX, (void (*)(void))name##_newctx },                \
+    { OSSL_FUNC_DIGEST_INIT, (void (*)(void))init },                           \
+    { OSSL_FUNC_DIGEST_UPDATE, (void (*)(void))upd },                          \
+    { OSSL_FUNC_DIGEST_FINAL, (void (*)(void))name##_internal_final },         \
+    { OSSL_FUNC_DIGEST_FREECTX, (void (*)(void))name##_freectx },              \
+    { OSSL_FUNC_DIGEST_DUPCTX, (void (*)(void))name##_dupctx },                \
+    OSSL_DISPATCH_FUNC_DIGEST_GET_PARAMS(name)
+
+# define OSSL_DISPATCH_FUNC_DIGEST_CONSTRUCT_END                               \
+    { 0, NULL }                                                                \
+};
+
+# define OSSL_FUNC_DIGEST_CONSTRUCT_START(name, CTX, blksize, dgstsize, flags, \
+                                          init, upd, fin)                      \
+OSSL_FUNC_DIGEST_ALLOC_METHODS(name, CTX)                                      \
+OSSL_FUNC_DIGEST_SET_FINAL(name, dgstsize, fin)                                \
+OSSL_FUNC_DIGEST_GET_PARAM(name, blksize, dgstsize, flags)                     \
+OSSL_DISPATCH_FUNC_DIGEST_COMMON(name, init, upd)
+
+# define IMPLEMENT_digest_functions_with_settable_ctx(name, CTX, blksize,      \
+                                                      dgstsize, flags,         \
+                                                      init, upd, fin,          \
+                                                      settable_ctx_params,     \
+                                                      set_ctx_params)          \
+OSSL_FUNC_DIGEST_CONSTRUCT_START(name, CTX, blksize, dgstsize, flags,          \
+                                 init, upd, fin),                              \
+    { OSSL_FUNC_DIGEST_SETTABLE_CTX_PARAMS,                                    \
+      (void (*)(void))settable_ctx_params },                                   \
+    { OSSL_FUNC_DIGEST_SET_CTX_PARAMS, (void (*)(void))set_ctx_params },       \
+OSSL_DISPATCH_FUNC_DIGEST_CONSTRUCT_END
+
+# define IMPLEMENT_digest_functions(name, CTX, blksize, dgstsize, flags,       \
+                                    init, upd, fin)                            \
+OSSL_FUNC_DIGEST_CONSTRUCT_START(name, CTX, blksize, dgstsize, flags,          \
+                                 init, upd, fin),                              \
+OSSL_DISPATCH_FUNC_DIGEST_CONSTRUCT_END
+
+const OSSL_PARAM *digest_default_gettable_params(void);
+int digest_default_get_params(OSSL_PARAM params[], int blksz, int paramsz,
+                              unsigned long flags);
+
+# ifdef __cplusplus
+}
+# endif
+
+#endif /* OSSL_DIGESTCOMMON_H */
```

Comment:
Regarding names, I wonder if `OSSL_FUNC_*` and `OSSL_DISPATCH_*` should be renamed to something a little more provider specific. After all, these are macros we don't publish anywhere...

Reply:
renamed and collapsed some macros that were not shared.
Also updated the docs.